dbms/src/Storages/DeltaMerge/ColumnFile/ColumnFileSchema.h

```diff
@@ -0,0 +1,140 @@
+    }
+};
+} // namespace std
+
+namespace DB
+{
+namespace DM
+{
+using Digest = UInt256;
+class ColumnFileSchema
+{
+private:
+    Block schema;
+
+    using ColIdToOffset = std::unordered_map<ColId, size_t>;
+    ColIdToOffset colid_to_offset;
+
+public:
+    explicit ColumnFileSchema(const Block & schema_)
+        : schema(schema_)
+    {
+        for (size_t i = 0; i < schema.columns(); ++i)
+            colid_to_offset.emplace(schema.getByPosition(i).column_id, i);
+    }
+
+    const DataTypePtr & getDataType(ColId column_id) const
+    {
+        /// Returns the data type of a column.
+        /// The specified column id must exist, otherwise something unexpected will happen.
+        auto index = colid_to_offset.at(column_id);
+        return schema.getByPosition(index).type;
+    }
+
+    String toString() const
+    {
+        return "{schema:" + (schema ? schema.dumpStructure() : "none") + "}";
+    }
+
+    const Block & getSchema() const { return schema; }
+    const ColIdToOffset & getColIdToOffset() const { return colid_to_offset; }
+};
+
+using ColumnFileSchemaPtr = std::shared_ptr<ColumnFileSchema>;
+
+class SharedBlockSchemas
+{
+private:
+    // we use sha256 to generate Digest for each ColumnFileSchema as the key of column_file_schemas,
+    // to minimize the possibility of two different schemas having the same key in column_file_schemas.
+    // Besides, we use weak_ptr to ensure we can remove the ColumnFileSchema,
+    // when no one use it, to avoid too much memory usage.
+    std::unordered_map<Digest, std::weak_ptr<ColumnFileSchema>> column_file_schemas;
+    std::mutex mutex;
+    BackgroundProcessingPool::TaskHandle handle;
+    BackgroundProcessingPool & background_pool;
+
+public:
+    explicit SharedBlockSchemas(DB::Context & context)
+        : background_pool(context.getBackgroundPool())
+    {
+        handle = background_pool.addTask([&, this] {
+            std::lock_guard<std::mutex> lock(mutex);
+            for (auto iter = column_file_schemas.begin(); iter != column_file_schemas.end();)
+            {
+                if (iter->second.expired())
+                {
+                    iter = column_file_schemas.erase(iter);
+                }
+                else
+                {
+                    ++iter;
+                }
+            }
+            return true;
+        },
+                                         false);
+    }
+
+    ~SharedBlockSchemas()
+    {
+        background_pool.removeTask(handle);
+    }
+
+    ColumnFileSchemaPtr find(const Digest & digest)
+    {
+        std::lock_guard<std::mutex> lock(mutex);
+        auto it = column_file_schemas.find(digest);
+        if (it == column_file_schemas.end())
+            return nullptr;
+        return it->second.lock();
+    }
+
+    void insert(const Digest & digest, const ColumnFileSchemaPtr & schema)
+    {
+        std::lock_guard<std::mutex> lock(mutex);
+        column_file_schemas.emplace(digest, schema);
+    }
```

Comment:
Better make it a `ColumnFileSchemaPtr getOrCreate(const Digest & digest, const Block & block)` inside one lock_guard